File: cirkill.py
import math
import sys
def solve():
        t= int(input())
                for ll in range(t):
                    cor=[]
                    a=[]
                    n= int(input())
                    c=0
                    for i in range(0,n):
                        l2=[]
                        for j in range(0,n):
                            l2.append(0)
                        a.append(l2)    
                    for i in range(n):
                        x= input().split()
                        l2=[]
                        l2.append( (int) (x[0]) )
                        l2.append((int) (x[1]))
                        cor.append(l2)
                        
                    for i in range(0,n):
                        for j in range(i+1,n):
                            x=  math.sqrt( (cor[j][0]-cor[i][0])*(cor[j][0]-cor[i][0])+ (cor[j][1]-cor[i][1])*(cor[j][1]-cor[i][1]) )
                            a[j][i]= a[i][j]=x
                    cnt=0
                    for i in range(0,n):
                        for j in range(i+1,n):
                            for k in range(j+1,n):
                                cnt=cnt+1
                                if(a[i][j]!=0 and  a[i][k]!=0 and a[j][k]!=0):
                                    if( (a[i][j]+ a[j][k]> a[i][k]) and ( a[i][k] + a[k][j] >a[i][j]) and ( a[i][k] + a[i][j] >a[j][k]) ):
                                        x= ( (cor[i][0]*cor[i][0]+ cor[i][1]*cor[i][1])* ( cor[j][1]- cor[k][1]) + ( cor[j][0]*cor[j][0] + cor[j][1]*cor[j][1])*( cor[k][1]-cor[i][1]) + (cor[k][0]* cor[k][0] + cor[k][1]*cor[k][1] )*(cor[i][1]- cor[j][1]) )
                                        y=  ( (cor[i][0]*cor[i][0]+ cor[i][1]*cor[i][1])* ( cor[k][0]- cor[j][0]) + ( cor[j][0]*cor[j][0] + cor[j][1]*cor[j][1])*( cor[i][0]-cor[k][0]) + (cor[k][0]* cor[k][0] + cor[k][1]*cor[k][1] )*(cor[j][0]- cor[i][0]) )
                                        z= 2*( cor[i][0]*( cor[j][1]- cor[k][1]) + cor[j][0]*( cor[k][1]- cor[i][1]) + cor[k][0]*( cor[i][1]- cor[j][1]) )
                                        R= math.sqrt( (cor[i][0]- x/z)*(cor[i][0]- x/z) + (cor[i][1]- y/z)*(cor[i][1]- y/z) )
                                        for m in range(n):
                                            dis_centre_ash = math.sqrt( (cor[m][0]-x/z)*(cor[m][0]-x/z) + (cor[m][1]-y/z)*(cor[m][1]-y/z) )
                                            if( dis_centre_ash<=R):
                                                c=c+1
                                        c=c-3
                    ans= c/(cnt*(n-3))
                    print (ans)
def main():
    solve()
    return 0

File: cirkill.cpp
#include<stdio.h>
#include<math.h>
#include<iostream>
#include<algorithm>
#include<vector>
#include<map>
#include<queue>
#include<string.h>
#define LL long long
#define maxm 10000000
using namespace std;


int det(int a1,int a2,int a3,int a4, int a5, int a6, int a7, int a8, int a9)
{
    int det_val= a1*(a5*a9 - a8*a6) - a2*(a4*a9-a6*a7)+ a3*(a4*a8-a5*a7);
    return det_val;
}
main()
{
    int c,n,i,j,k,m,fourth,first,second,third,val;
    double ans=0.0;
    int cor[30][2];
    int test_cases;
    scanf("%d",&test_cases);
    while(test_cases--)
    {
        c=0;
        scanf("%d",&n);
        for(i=0;i<n;i++)
        {
            scanf("%d %d",&cor[i][0],&cor[i][1]);
        }


        int cnt=0;
        for(i=0;i<n-2;i++)
        {
            for(j=i+1;j<n-1;j++)
            {
                for(k=j+1;k<n;k++)
                {
                    cnt++;
                   int area= abs(cor[i][0]*(cor[j][1]-cor[k][1]) + cor[j][0]*(cor[k][1]-cor[i][1]) + cor[k][0]*(cor[i][1]-cor[j][1]));

                        if( area!=0 )// forms a circle
                        {

                                for(m=0;m<n;m++)
                                {
                                    if(m==i || m==j || m==k) continue;
                                    else
                                    {
                                        first = (cor[m][0]*cor[m][0]+ cor[m][1]*cor[m][1])* det(cor[i][0],cor[i][1],1,cor[j][0],cor[j][1],1,cor[k][0],cor[k][1],1);
                                        second= cor[m][0]* det((cor[i][0]*cor[i][0]+ cor[i][1]*cor[i][1]), cor[i][1],1, (cor[j][0]*cor[j][0]+ cor[j][1]*cor[j][1]), cor[j][1],1, (cor[k][0]*cor[k][0]+ cor[k][1]*cor[k][1]), cor[k][1],1);
                                        third= cor[m][1]*det((cor[i][0]*cor[i][0]+ cor[i][1]*cor[i][1]), cor[i][0],1, (cor[j][0]*cor[j][0]+ cor[j][1]*cor[j][1]), cor[j][0],1, (cor[k][0]*cor[k][0]+ cor[k][1]*cor[k][1]), cor[k][0],1);
                                        fourth= det((cor[i][0]*cor[i][0]+ cor[i][1]*cor[i][1]), cor[i][0],cor[i][1], (cor[j][0]*cor[j][0]+ cor[j][1]*cor[j][1]),cor[j][0], cor[j][1], (cor[k][0]*cor[k][0]+ cor[k][1]*cor[k][1]), cor[k][0],cor[k][1]);

                                        val= first- second+third-fourth;
                                        if(val<=0)
                                            c++;
                                    }
                                }


                    }
                }
            }
        }

printf("total number killed places=%d  and total place = %d\n",cnt*(n-3),c);
        ans=  (double)c/ (double)(cnt*(n-3));
        printf("%.6lf\n",ans);

    }
return 0;
}
